cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -275,56 +290,170 @@ void accumulate_vertex_results(
+        [] __device__(auto pair) { 
+          return thrust::make_tuple(thrust::get<1>(pair), weight_t{0}); 
+        });
+      
+      // Create key_bucket_t from the frontier vertices
+      key_bucket_t<vertex_t, void, multi_gpu, true> vertex_list(handle);
+      vertex_list.insert(reusable_vertex_buffer.begin(), reusable_vertex_buffer.begin() + frontier_count);
+
+      // Compute deltas for frontier vertices
+      per_v_transform_reduce_outgoing_e(
+        handle,
+        graph_view,
+        vertex_list,
+        src_properties.view(),
+        dst_properties.view(),
+        cugraph::edge_dummy_property_t{}.view(),
+        [d] __device__(auto, auto, auto src_props, auto dst_props, auto) {
+          if (thrust::get<0>(dst_props) == d) {
+            auto sigma_v = static_cast<weight_t>(thrust::get<1>(src_props));
+            auto sigma_w = static_cast<weight_t>(thrust::get<1>(dst_props));
+            auto delta_w = thrust::get<2>(dst_props);
+            
+            return (sigma_v / sigma_w) * (1 + delta_w);
+          } else {
+            return weight_t{0};
+          }
+        },
+        weight_t{0},
+        reduce_op::plus<weight_t>{},
+        reusable_delta_buffer.begin(),
+        do_expensive_check);
+      
+      // Scatter deltas back to main deltas array
+      thrust::scatter(
+        handle.get_thrust_policy(),
+        reusable_delta_buffer.begin(),
+        reusable_delta_buffer.begin() + frontier_count,
+        reusable_vertex_buffer.begin(),
+        deltas.begin()
+      );
+      
+      // Combined operation: update properties and accumulate centralities in one kernel
+      
+      // Get the property iterators outside the lambda (host-side)
+      auto src_props_first = src_properties.mutable_view().major_value_firsts()[0];  // Source properties are major
+      auto dst_props_first = dst_properties.mutable_view().minor_value_first();      // Destination properties are minor
+      
+      thrust::for_each(
+        handle.get_thrust_policy(),
+        thrust::make_zip_iterator(reusable_vertex_buffer.begin(), reusable_delta_buffer.begin()),
+        thrust::make_zip_iterator(reusable_vertex_buffer.begin(), reusable_delta_buffer.begin()) + frontier_count,
+        [distances = distances.begin(),
+         sigmas = sigmas.begin(),
+         deltas = deltas.begin(),
+         centralities = centralities.data(),
+         src_props_first,
+         dst_props_first,
+         d] __device__(auto pair) {
+          auto v = thrust::get<0>(pair);
+          auto delta = thrust::get<1>(pair);
+          
+          // Update source properties (major)
+          src_props_first[v] = thrust::make_tuple(distances[v], sigmas[v], deltas[v]);
+          
+          // Update destination properties (minor)
+          dst_props_first[v] = thrust::make_tuple(distances[v], sigmas[v], deltas[v]);
+          
+          // Accumulate centralities
+          centralities[v] += delta;
+        });
```

Comment:
Don't do this. This code won't work in multi-GPU.

Call thrust::for_each to update just centralities.

To update edge properties, you just need to update deltas for the vertices in vertex_list. No need to read/write distances & sigmas.

Call update_edge_src|dst_property with vertex_list. To do this, you may create a separate edge_src|dst_property objects for distances & sigmas and deltas.

When you need all three, you can use view_concat (https://github.com/rapidsai/cugraph/blob/branch-25.08/cpp/include/cugraph/edge_src_dst_property.hpp#L829).

Reply:
And later, for final last level tuning, we can even skip this for SG, but we will revisit this later.